src/node/oscillator.rs

```diff
@@ -0,0 +1,1006 @@
+                    (phase + 2. * PI) * (TABLE_LENGTH_F32 / (2.0 * PI))
+                } else {
+                    phase * (TABLE_LENGTH_F32 / 2.0 * PI)
+                }
+            })
+            .collect();
+
+        let incr_phases: Vec<f32> = cplxs
+            .iter()
+            .enumerate()
+            .map(|(idx, _)| TABLE_LENGTH_F32 * idx as f32 * (computed_freq / sample_rate))
+            .collect();
+
+        let interpol_ratios: Vec<f32> = incr_phases
+            .iter()
+            .map(|incr_phase| incr_phase - incr_phase.floor())
+            .collect();
+        let normalizer = if !disable_normalization {
+            let norm = Self::get_normalizer(
+                phases.clone(),
+                incr_phases.clone(),
+                interpol_ratios.clone(),
+                norms.clone(),
+                computed_freq,
+            );
+            Some(norm)
+        } else {
+            None
+        };
+
+        Self {
+            computed_freq,
+            sample_rate,
+            cplxs,
+            norms,
+            phases,
+            incr_phases,
+            interpol_ratios,
+            normalizer,
+        }
+    }
+
+    fn get_normalizer(
+        mut phases: Vec<f32>,
+        incr_phases: Vec<f32>,
+        interpol_ratios: Vec<f32>,
+        norms: Vec<f32>,
+        computed_freq: f32,
+    ) -> f32 {
+        let mut samples: Vec<f32> = Vec::new();
+
+        if computed_freq == 0. {
+            return 1.;
+        }
+
+        while phases[1] <= TABLE_LENGTH_F32 {
+            let mut sample = 0.0;
+            for i in 1..phases.len() {
+                let gain = norms[i];
+                let phase = phases[i];
+                let incr_phase = incr_phases[i];
+                let mu = interpol_ratios[i];
+                let idx = (phase + incr_phase) as usize;
+                let inf_idx = idx % TABLE_LENGTH_USIZE;
+                let sup_idx = (idx + 1) % TABLE_LENGTH_USIZE;
+                // Linear interpolation
+                sample += (SINETABLE[inf_idx] * (1. - mu) + SINETABLE[sup_idx] * mu) * gain;
+                phases[i] = phase + incr_phase;
+            }
+            samples.push(sample);
+        }
+
+        1. / samples
+            .iter()
+            .copied()
+            .reduce(f32::max)
+            .expect("Maximum value not found")
+    }
+
+    fn compute_params(&mut self, new_comp_freq: f32) {
+        // No need to compute if frequency has not changed
+        if (self.computed_freq - new_comp_freq).abs() < 0.01 {
+            return;
+        }
+
+        for incr_phase in &mut self.incr_phases {
+            *incr_phase *= new_comp_freq / self.computed_freq;
+        }
+
+        for (r, incr_ph) in self.interpol_ratios.iter_mut().zip(self.incr_phases.iter()) {
+            *r = incr_ph - incr_ph.floor();
+        }
+
+        self.computed_freq = new_comp_freq;
+    }
+
+    fn set_periodic_wave(&mut self, periodic_wave: PeriodicWave) {
```

Comment:
Ideally, these computations (and allocations) could be done on the control thread.
You could extend the `OscMsg` to include the pre-calculated fields